### PR TITLE
Add portus chart

### PR DESCRIPTION
**Pull request by rendhalver**
This is a resubmit of #1284.
Thanks to @so0k for all the hard work!
This is currently the same as the last commit in that PR.

It looks like there are a few differences in layout since that PR was submitted so I will try to address those concerns as I go.
There were also a few concerns not address in the original PR so I will address those as well.

### CODE REVIEW

**Comment by unguiculus** [comment]
Please check our best practices, especially those on naming resources and standard labels. Resource metadata should generally look like so:

```yaml
name: {{ template "portus.fullname" . }}
labels:
  heritage: {{ .Release.Service }}
  release: {{ .Release.Name }}
  chart: {{ .Chart.Name }}-{{ .Chart.Version }}
  app: {{ template "portus.name" . }}
```

**Comment by rendhalver** [comment]
Thanks @unguiculus I will get those fixed.

**Comment by rendhalver** [comment]
@unguiculus This Chart has multiple components but there isn't a definition $component.name in the _helpers.tpl file and I am not sure of the proper way to do that.
Can you point me to some docs on how to achieve that?

On a side note I am very tempted to split out the registry component into it's own chart and then add that as a requirement for this chart so the templates are a bit simpler.

I am also new to writing Charts so I am still learning how best to do them.

**Comment by unguiculus** [comment]
There's a chart for Docker registry already: https://github.com/kubernetes/charts/tree/master/incubator/docker-registry

**Comment by rendhalver** [comment]
True. I forgot about that Chart.
I had already discounted that because doesn't have the options we need to setup the registry for Portus.
Maybe I can expand that one so it can be configured with those extra options.

As for my other question what is the best way to define those $component.name variables?

**Comment by rendhalver** [comment]
@unguiculus I think worked out what I needed to change.
The latest commit addresses the label inconsistencies.

**Comment by unguiculus** [comment]
Please create a PR for the registry chart to add the stuff you need. We should also see that we can move it to stable so you can use it.

**Comment by rendhalver** [comment]
Will do.
Thanks for the help @unguiculus It is much appreciated. :)

**Comment by so0k** [comment]
The reason I had them bundled was the auth token service configuration required on the registry side.

If the registry chart exposes that, it should be fine to define it as a dependency though

**Comment by rendhalver** [comment]
@unguiculus What else needs changing in the docker-registry chart to move it to stable?

File: stable/portus/Chart.yaml
Comment (unguiculus) [inline stable/portus/Chart.yaml]: Add `appVersion`

File: stable/portus/Chart.yaml
Comment (unguiculus) [inline stable/portus/Chart.yaml]: Use Github username

**Comment by rendhalver** [comment]
@viglesiasce This is dependent on my other PR for getting docker-registry usable and put of incubator #2801. Once that is sorted I can update this to reflect the new location of docker-registry.

**Comment by unguiculus** [comment]
I think the Docker registry chart looks pretty good. I'd suggest you create a PR to move it to stable.

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by rendhalver** [comment]
It needs to be redone or rebased. Redoing it seems quicker at this stage. #2612 got merged that changed a lot of the docker-registry chart.
I will get that sorted when I get a chance.

**Comment by rendhalver** [comment]
The docker-registry PR has been redone in #2926.

**Comment by rendhalver** [comment]
@unguiculus This is my PR to promote docker-registry to stable: #2944
I can update the requirements on this once that is merged.

File: stable/portus/Chart.yaml
Comment (so0k) [reply stable/portus/Chart.yaml]: can remove me if this is holding back the merge :)

**Comment by unguiculus** [comment]
Let's try it again. Timeouts have been increased.

**Comment by rendhalver** [comment]
I ran a test locally and it's failing with default settings because it expects a tls secret setup with the cert for docker-registry.
It's listed in the README.

**Comment by unguiculus** [comment]
You'll have to change it such that CI passes. Otherwise it can't be merged.

**Comment by bonhamjp** [comment]
Hello @rendhalver the work on this chart is looking very good! 

I wasn't aware that anyone was working on implementing a Portus chart, until I came across this. I work at SUSE, and we have also been developing a Portus chart. Our work can be found at https://github.com/kubic-project/caasp-services/tree/master/contrib/helm-charts/portus

Our goal was to propose our chart here when we finished developing it. Would you be interested in collaborating?

**Comment by rendhalver** [comment]
Hi @bonhamjp I wasn't aware SUSE was working on a Chart for Portus so I guess we both learned something. :)
Collaboration sounds good to me.

Are you using your Chart in production?

The portus parts of your chart look similar to mine and has some good features that would help get this one finished like generating a tls cert in the container.

I have tried to make this chart use other charts to deploy the components (docker-registry for example).
I am in k8s Slack as rendhalver as well if you want to chat about it there.

**Comment by so0k** [comment]
@bonhamjp - I pinged portus maintainers on the portus issues related to k8s deployments almost a year ago (when I submitted the first chart proposal - mostly based on the work from Nordstorm)

However, I've noticed Portus 2.2 release branche was broken for several months and overall development seems stuck - which is why we no longer run Portus at Honestbee

Happy to see work is still ongoing

**Comment by rendhalver** [comment]
@so0k I couldn't get 2.2 working in our setup either.
I will downgrade it in this chart till it does work.

**Comment by so0k** [comment]
you're supposed to use an image tagged with a conference name, main reason I dropped Portus

**Comment by rendhalver** [comment]
@so0k What do you mean by "image tagged with a conference name"?

**Comment by so0k** [comment]
https://github.com/SUSE/Portus/issues/1418#issuecomment-355291208

**Comment by rendhalver** [comment]
Thanks @so0k. I will keep an eve on that.

**Comment by rendhalver** [comment]
@unguiculus I just rebased this to pickup the new tests and It looks like we have test success now.
Do you have any further concerns about this one?

**Comment by rendhalver** [comment]
I just realised this will need changes to work with Portus 2.3.
I will get those added as soon as I can.

**Comment by flavio** [comment]
I think work on this PR can be resumed. @mssola released a new version of the Portus image that fixes the issue reported by @rendhalver

**Comment by rendhalver** [comment]
I noticed that new version had been released.
Thanks for the reminder @flavio.
I will try and requisition some time to get this working.

**Comment by rendhalver** [comment]
@unguiculus Do you have any further concerns?
I managed to get this working with the new 2.3 release.

**Comment by davidkarlsen** [comment]
Test seem to fail:
```
I0814 18:31:38.332] Installing chart 'stable/portus' into namespace 'presubmit-2766-12568-wpavpmqpitulekn4'...
I0814 18:31:38.332] Chart does not provide test values. Using defaults...
W0814 18:31:39.426] Error: release portus-wpavpmqpitulekn4 failed: Deployment.apps "portus-wpavpmqpitulekn4" is invalid: spec.template.spec.containers[1].volumeMounts[0].name: Not found: "secrets"
I0814 18:31:40.107] Waiting for rollout to finish: 0 of 2 updated replicas are available...
```

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/helm_charts/2766/pull-charts-e2e/12568/

**Comment by rendhalver** [comment]
Thanks I will get that fixed.

**Comment by rendhalver** [comment]
Why didn't that failure fail the build?

**Comment by davidkarlsen** [comment]
@rendhalver Have a look at https://github.com/helm/chart-testing/issues/9

**Comment by rendhalver** [comment]
I see what's going on.
It's the problem we had with testing before where the deploy is assuming the Secret exists before the chart it setup.
I will put in the code to create those certs if they don't exist.

File: stable/portus/templates/portus-post-delete-job.yaml
Comment (unguiculus) [inline stable/portus/templates/portus-post-delete-job.yaml:19]: Add `release` label to selector.

File: stable/portus/templates/minio-create-bucket-job.yaml
Comment (unguiculus) [inline stable/portus/templates/minio-create-bucket-job.yaml]: Add `release` label to selector.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/portus/templates/portus-post-delete-job.yaml
Comment (rendhalver) [reply stable/portus/templates/portus-post-delete-job.yaml:19]: Thanks for the reminder.
I will get that fixed.

File: stable/portus/templates/minio-create-bucket-job.yaml
Comment (rendhalver) [reply stable/portus/templates/minio-create-bucket-job.yaml]: Thanks for the reminder.
I will get that fixed.

**Comment by unguiculus** [comment]
With the latest chart-testing fix the CI job finally fails as expected.

File: stable/portus/.helmignore
Comment (giacope) [inline stable/portus/.helmignore:21]: Please add the OWNERS file in the .helmignore - https://github.com/helm/charts#owning-and-maintaining-a-chart

**Comment by flavio** [comment]
Triggering the reopening of this issue. @mssola is going to take care of that unless @rendhalver or others have something against it.

**Comment by rendhalver** [comment]
I am happy to help get it working. I haven't had the time to devote to it lately.
I am also happy to be a maintainer and OWNER if you need one.

**Comment by rendhalver** [comment]
And thanks. :)
I had a few ideas on how to get it working but didn't get time top test them.
Ping my on your new PR and I will add those thoughts there.

**Pull request closed without merging**
